### PR TITLE
Improve validation #1263 #1255

### LINKED ISSUE #1263: Revisit if `types` is indeed an illegal name
**Minimal .gitlab-ci.yml illustrating the issue**
```yml
---
types:
  script:
    - echo "test"

test:
  needs: [types]
  script:
    - echo "test"
```

**Expected behavior**
According to the documentation, it should not work.
https://docs.gitlab.com/ee/ci/jobs/#job-name-limitations

However, it does work:
https://gitlab.com/hverlin/gitlab-ci-local-tests/-/pipelines/1324968259

Probably this is a limitation that they forgot to remove from the documentation?

```shell
> gitlab-ci-local
needs: [types] for test could not be found
```

Two solutions:
- remove `types` from: https://github.com/firecow/gitlab-ci-local/blob/ba3509a42b5e0155913fb09d0885c93284df1f39/src/job.ts#L69-L73
- add a warning to the console if a job name matches an invalid name

### PR DESCRIPTION
**Pull request by hverlin**
Fix #1263 #1255

Given 
```yml
test:
  needs: [types]
  when: 'manual2'
  artifacts:
    reports:
      junit: 3
  script: ''
  data: "test"
```

```
Invalid .gitlab-ci.yml configuration
        • 'data' property is not expected to be here at test
        • property 'script' must not have fewer than 1 characters at test.script
        • 'when' property must be one of [on_success, on_failure, always, never, manual, delayed] (found manual2) at test.when
        • 'junit' property type must be string at test.artifacts.reports.junit
```
image

### CODE REVIEW

**Comment by hverlin** [comment]
@firecow @ANGkeith Let me know if these changes are ok. I will then add tests.

- Regarding schema validation, output was done directly using `console`/`assert` before. Should I switch to `writeStreams`?
- I am now always writing the expanded pipeline `yaml` under the `.gitlab-ci-local`. I think this is useful outside validation in case you want to inspect what was parsed without having to always do `gcl --preview > /tmp/my-pipeline.yml`

**Comment by ANGkeith** [comment]
looks good to me

**Comment by hverlin** [comment]
@firecow Anything you would change here before I add tests?

**Comment by firecow** [comment]
@hverlin Nah, nothing much. All looks good.


```
Validation errors:
        test must NOT have additional properties
```

Is it possible to print exactly what properties are additional?

**Comment by hverlin** [comment]
I reused code from https://github.com/apideck-libraries/better-ajv-errors, and I am now getting better error message.

Given 
```yml
test:
  needs: [types]
  when: 'manual2'
  artifacts:
    reports:
      junit: 3
  script: ''
  data: "test"
```

```
Invalid .gitlab-ci.yml configuration
        • 'data' property is not expected to be here at test
        • property 'script' must not have fewer than 1 characters at test.script
        • 'when' property must be one of [on_success, on_failure, always, never, manual, delayed] (found manual2) at test.when
        • 'junit' property type must be string at test.artifacts.reports.junit
```
image

File: src/index.ts
Comment (hverlin) [inline src/index.ts:33]: This makes it possible to run  `gitlab-ci-local` with `bun` to that it's faster to iterate
```
bun ../../gitlab-ci-local/src/index.ts
```

File: src/schema-error.ts
Comment (hverlin) [inline src/schema-error.ts:1]: I did not use the code directly as:
- I was getting NPE for some cases
- I am not sure if it makes sense to bring the other libraries that they use for suggestions
- We can improve some error messages later if needed

Comment (firecow) [reply src/schema-error.ts:1]: I don't particularly like this, but I'll let it slip :smile:

**Review by firecow — APPROVED** [review]

**Comment by firecow** [comment]
@ANGkeith I'll let you review this as well, before I'm merging.

**Comment by ANGkeith** [comment]
very nice

File: src/validator.ts
Comment (hverlin) [reply src/validator.ts]: Updated it with link support
image

**Review by ANGkeith — APPROVED** [review]

**Pull request merged**
